OrbitQt/AccessibilityAdapter.cpp

```diff
@@ -0,0 +1,269 @@
+ * exists OrbitGlWidgetAccessible. This class inherits from QAccessibleWidget and uses its default
+ * functionality to walk *up* the tree, but replaces all methods to walk *down* the tree and
+ * forwards those calls to the AccessibleInterface of GlCanvas. It is important to note that the
+ * *parent* methods of the AccessibleInterface associated with GlCanvas are thus never invoked.
+ *
+ *
+ *                OrbitGl                    +                                 OrbitQt
+ *                                           |
+ *                                           |                          ^
+ *                                           |                          | parent(): QT Default
+ *                                           |                          |
+ *                                           |  +-------------+   +-----+----------------------+
+ *                                           |  |OrbitGlWidget+<--+OrbitGlWidgetAccessible     |
+ *                                           |  +-------------+   +-----+----------------------+
+ *                                           |                          |
+ *                                           |                          |
+ * +-----------+   +----------------------+  |                          |  Adapter for child()
+ * | GlCanvas  +<--+AccessibleInterface +<----------------------------+
+ * +-----------+   +--+-------------+-----+  |
+ *                    |             ^ Parent |
+ *                    v Child(i)    |        |
+ * +-----------+   +--+-------------+-----+  |                    +-----------------------------+
+ * | TimeGraph +<--+AccessibleInterface +<----------------------+AccessibilityAdapter         |
+ * +-----------+   +----------------------+  |                    +-----------------------------+
+ *                                           |
+ *                                           +
+ *
+ * The OrbitGlWidgetAccessible is automatically created when an OrbitGlWidget is constructed by
+ * installing a QT Accessibility Factory (InstallAccessibilityFactories()).
+ *
+ * To make sure adapters created by AccessibilityAdapter::GetOrCreateAdapter are deleted when the
+ * corresponding AccessibleInterface is deleted, these interfaces register themselves in the
+ * AccessibleInterfaceRegistry (see OrbitGlAccessibility.h), which in turn allows to register a
+ * callback on interface deletion. AccessibilityAdapter registers itself for this callback and
+ * cleans up interfaces accordingly.
+ */
+
+#include "AccessibilityAdapter.h"
+
+#include <QWidget>
+#include <mutex>
+
+#include "AccessibleInterfaceRegistry.h"
+
+using orbit_accessibility::AccessibleInterface;
+using orbit_accessibility::AccessibleInterfaceRegistry;
+
+namespace orbit_qt {
+
+absl::flat_hash_map<const AccessibleInterface*, QAccessibleInterface*>
+    AccessibilityAdapter::interface_map_;
+absl::flat_hash_set<const QAccessibleInterface*> AccessibilityAdapter::managed_adapters_;
```

Comment:
If I understand correctly, this should contain `unique_ptr`s, and the `new` and `delete` below should be replaces with the usage of `unique_ptr`. Or is there some specific problem with types? Because of the protected destructor? But why can't this contain pointers to `AccessibilityAdapter` if we know that's the type of the elements?

Reply:
The reason was that I needed to be able to perform a O(1) lookup based on the raw pointer (`AccessibilityAdapter::OnInterfaceDeleted`), so I can't store unique_ptr in this set. But I agree that we should avoid raw new and delete. This actually pointed me to a (IMO) better solution now: `managed_adapters_` is now a map from `AccessibleInterface` to `std::unique_ptr<AccessibilityAdapter>`. This map then actually owns the memory of all adapters created through `GetOrCreateAdapter`, and `interface_map_` contains the raw pointers for these and all interfaces managed elsewhere.